examples/async2-passthru.rs

```diff
@@ -0,0 +1,447 @@
+use clap::Parser;
+use etherparse::{InternetSlice::*, LinkSlice::*, TransportSlice::*, *};
+use futures::{task::AtomicWaker, Future};
+use ndisapi::{EthRequest, FilterFlags};
+use std::{
+    ffi::c_void,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+use tokio::{runtime::Builder, sync::oneshot};
+use windows::{
+    core::Result,
+    Win32::{
+        Foundation::{GetLastError, CloseHandle, BOOLEAN, HANDLE},
+        System::Threading::{
+            CreateEventW, RegisterWaitForSingleObject, ResetEvent, UnregisterWaitEx, INFINITE,
+            WT_EXECUTEINWAITTHREAD,
+        },
+    },
+};
+
+// The struct NdisapiAdapter represents a network adapter with its associated driver and relevant handles.
+pub struct NdisapiAdapter {
+    driver: Arc<ndisapi::Ndisapi>, // The network driver for the adapter.
+    adapter_handle: HANDLE,        // The handle of the network adapter.
+    notif: Win32EventFuture,       // A future that resolves when a Win32 event is signaled.
+}
+
+impl NdisapiAdapter {
+    /// Create NdisapiAdapter
+    pub fn new(
+        driver: Arc<ndisapi::Ndisapi>, // The network driver for the adapter.
+        adapter_handle: HANDLE,        // The handle of the network adapter.
+        flags: ndisapi::FilterFlags,   // The filter flags for the adapter.
+    ) -> Result<Self> {
+        let event_handle = unsafe {
+            // Creating a Win32 event without a name. The event is manual-reset and initially non-signaled.
+            CreateEventW(None, true, false, None)?
+        };
+
+        // Setting the event for packet capture for the specified adapter.
+        driver.set_packet_event(adapter_handle, event_handle)?;
+
+        // Setting the operating mode for the specified adapter.
+        driver.set_adapter_mode(adapter_handle, flags)?;
+
+        Ok(Self {
+            adapter_handle,
+            driver,
+            notif: Win32EventFuture::new(event_handle)?, // Creating a new Win32EventFuture with the event handle.
+        })
+    }
+
+    /// Wait for a packet event to be signaled before continuing with the packet capture process.
+    async fn wait_for_packet(&mut self) -> Result<()> {
+        Pin::new(&mut self.notif).await
+    }
+
+    /// Read a packet from the network adapter and return it as an `EthRequest` struct.
+    pub async fn read_packet(&mut self, packet: &mut EthRequest) -> Result<()> {
+        let driver = self.driver.clone();
+
+        // first try to read packet
+        if unsafe { driver.read_packet(packet) }.is_ok() {
+            return Ok(());
+        }
+
+        let result = self.wait_for_packet().await; // wait for packet event
+        match result {
+            Ok(_) => {
+                if unsafe { driver.read_packet(packet) }.ok().is_some() {
+                    Ok(())
+                } else {
+                    Err(unsafe { GetLastError() }.into())
+                }
```

Comment:
Can we extract this bit into a helper?

Reply:
I attempted the suggested change, but in my opinion, it doesn't enhance the readability.